File: nesting/src/main/java/brackets/nesting/BracketTreeBuilder.java
package brackets.nesting;

import java.util.Vector;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import brackets.BracketEnthusiast;

/** The Builder for a Tree of Bracket Nodes.
 */
public final class BracketTreeBuilder {

	/** The Root Node in the Tree.
	 */
	private BracketNode mRoot = null;

	@Nullable
	BracketNode getRoot() {
		return mRoot;
	}

	/** Build a Top Level Array of Bracket Nodes from an Enthusiast.
	 * Clears the Builder before and after this operation.
	 * @param enthusiast The BracketEnthusiast to provide data.
	 * @return An Array of BracketNode which may be empty, or null if brackets are imbalanced.
	 */
	@Nullable
	public BracketNode[] buildFromEnthusiast(
		@Nonnull final BracketEnthusiast enthusiast
	) {
		// Clear Builder before starting
		mRoot = null;
		// Ensure that the Enthusiast contains balanced data
		if (!enthusiast.getAreBracketsBalanced())
			return null;
		// Obtain the Data from the Enthusiast
		final int[] pairs = enthusiast.getBracketPairs();
		if (pairs == null)
			return null;
		// Record the top Level Nodes
		var topLevelNodes = new Vector<BracketNode>(2, 2);
		// Parse the Pairs
		int index = 0;
		while (index < pairs.length) {
			// Try to add to the Root Node
			if (addNode(pairs[index], pairs[index + 1])) {
				index += 2; // Next Pair
			} else {
				topLevelNodes.add(mRoot); // Pop Root and Try Again
				mRoot = null;
			}
		}
		// There's always a root node remaining
		topLevelNodes.add(mRoot);
		mRoot = null;
		return topLevelNodes.toArray(new BracketNode[0]);
	}

	/** Add A Node to The Builder tree.
	 * Empty builder accepts any node, sets it to builder root.
	 * Builder with root tries to add the node anywhere within.
	 * Returns false if unable to add the nade within the root or it's sub-nodes.
	 * Modifies the recent member reference to point to the most recently added node.
	 * @param open  The index of the opening bracket.
	 * @param close The index of the closing bracket.
	 * @return Whether the node was successfully added to the builder.
	 */
	boolean addNode(
		final int open,
		final int close
	) {
		if (mRoot == null) {
			// The builder is empty
			mRoot = new BracketNode(open, close);
			return true;
		}
		// Non-empty tree
		return mRoot.addNode(open, close);
	}

	/** Reset the Builder references.
	 */
	void clear() {
		mRoot = null;
	}

}

File: nesting/src/main/java/brackets/nesting/BracketTreeBuilder.kt
package brackets.nesting

import brackets.BracketEnthusiast

/**
 */
class BracketTreeBuilder {

    private var root: BracketNode? = null

    private var recent: BracketNode? = null

    /** Add A Node to The tree.
     * @param
     * @param
     * @return
     */
    fun addNode(
        a: Int, b: Int
    ) : Boolean {
        if (root == null) {
            root = BracketNode(a, b)
            recent = null
            return true
        }
        //todo:

        return false
    }

    /** Build a Top Level Array of Bracket Nodes from an Enthusiast.
     * @param enthusiast The BracketEnthusiast to provide data.
     * @return An Array of BracketNode, or an empty Array if data invalid or empty.
     */
    fun buildFromEnthusiast(
        enthusiast: BracketEnthusiast,
    ) : Array<BracketNode> {
        // Ensure that the Enthusiast contains balanced data
        if (!enthusiast.areBracketsBalanced)
            return emptyArray()
        // Obtain the Data from the Enthusiast
        val pairs = enthusiast.getBracketPairs()
            ?: return emptyArray()
        // Clear the builder's root
        clear()
        // Convert Pairs into Nodes
        val topLevelNodes = ArrayList<BracketNode>()
        // Parse the Data Stream
        var index = 0
        while (index < pairs.size) {
            // Try to add to the Root Node
            if (addNode(pairs[index], pairs[index + 1])) {
                index += 2 // Next Pair
            } else {
                // Pop Root and Try Again
                topLevelNodes.add(root!!)
                clear()
            }
        }
        root?.also { topLevelNodes.add(it) }
        clear()
        return topLevelNodes.toTypedArray()
    }

    /** Reset the Builder references.
     */
    internal fun clear() {
        root = null
        recent = null
    }

}